guides/source/active_record_postgresql.md

```diff
@@ -588,7 +588,7 @@ ActiveRecord::Base.connection.transaction do
 end
 ```
 
-The `:deferrable` option can also be set to `true` or `:immediate`, which has the same effect. Both options let the foreign keys keep the default behavior of checking the constraint immediately, but allow manually deferring the checks using `SET CONSTRAINTS ALL DEFERRED` within a transaction. This will cause the foreign keys to be checked when the transaction is committed:
+When the `:deferrable` option was set to `:immediate`, let the foreign keys keep the default behavior of checking the constraint immediately, but allow manually deferring the checks using `SET CONSTRAINTS ALL DEFERRED` within a transaction. This will cause the foreign keys to be checked when the transaction is committed:
```

Comment:
I'm wondering if this statement is correct. It looks like saying that `NOT DEFERRABLE` and `DEFERRABLE INITIALLY IMMEDIATE` without `SET CONSTRAINTS ALL DEFERRED` behave the same way. 

However, when I tested the deferrable unique (not foreign key, though) constraint, these two options behaved differently.


[Example1 : No option given](https://gist.github.com/yahonda/1adcf1a776a385b8ec47c9cdf5dcf08b#example1--no-option-given) raised `ERROR:  duplicate key value violates unique constraint "example1_row_col_key"`. On the other hand [Example 5 : DEFERRABLE INITIALLY IMMEDIATE](https://gist.github.com/yahonda/1adcf1a776a385b8ec47c9cdf5dcf08b#example-5--deferrable-initially-immediate) the update successfully finished.

Reply:
I recall https://stackoverflow.com/questions/5300307/not-deferrable-versus-deferrable-initially-immediate explains it about unique key constraints, but I am not sure if it also applies to foreign key constraints.

Reply:
For unique constraints, we was confirmed that `NOT DEFERRABLE` performs `each row check` and `DEFERRABLE INITIALLY IMMEDIATE` performs each `statement check`.
Foreign key constraints and unique constraint internally create Trigger, so it is possible that they are handled similarly, but I have not yet read and understand the Postgres source code.
Since I am on vacation for a week, I will continue my investigation the week after next.

However, while unique keys behave differently for `each row` and `statement`, I haven't come up with a example query where a foreign key constraint would produce the same difference in behavior as a unique key constraint.
At the moment, I guess that checking at `each row` and at `each statement` will have the same result if the foreign key constraint is violated. Hopefully I'll come up with something examples while I'm on vacation.

Reply:
In the PostgreSQL specification, there is a description of compatibility differences with SQL-Standard.

> https://postgrespro.com/docs/postgresql/12/sql-createtable#id-1.9.3.85.9.4
> Non-Deferred Uniqueness Constraints
> When a UNIQUE or PRIMARY KEY constraint is not deferrable, PostgreSQL checks for uniqueness immediately whenever a row is inserted or modified. The SQL standard says that uniqueness should be enforced only at the end of the statement; this makes a difference when, for example, a single command updates multiple key values. To obtain standard-compliant behavior, declare the constraint as DEFERRABLE but not deferred (i.e., INITIALLY IMMEDIATE). Be aware that this can be significantly slower than immediate uniqueness checking.

- SQL-Standard checks constraints at the end of **each statement**.
- However, in PostgreSQL, when a `UNIQUE` or `PRIMARY KEY` constraint is `NOT DEFERRABLE`, it is checked with **each row** instead of **each statement**.
- When the `DEFERRABLE` option is given, check with **each statement** according to the SQL standard.

If we trust this statement, it seems that the difference in behavior between `NOT DEFERRABLE` and `DEFERRABLE` is limited to `CHECK` and `PRIMARY KEY` constraints.
Therefore, the behavior is the same between `NOT DEFERRABLE` and `DEFERRABLE` for `FOREIGN KEY` constraints.

Reply:
Wait... it says UNIQUE and PRIMARY, not CHECK.
I'm going to recheck the specification 😵

Reply:
@yahonda  I was reading PostgreSQL code for a while. It was very difficult for me. 🤯 

Im my current understanding is that the default behavior does not change for `FOREIGN KEY`.
Only `UNIQUE` and `PRIMARY` change the default behavior for `NOT DEFERRABLE` and `DEFERRABLE`.

Here are some 📝 

- `UNIQUE INDEX` is checked on insert/update in each Row to speed up processing
- In PostgreSQL, `UNIQUE` and `PRIMARY` constraints physically create a `UNIQUE INDEX`.
  - `UNIQUE/PRIMARY` constraints of `NOT DEFERRABLE` behave like regular `UNIQUE INDEX`.
- PostgreSQL can only create a `DEFERRABLE UNIQUE INDEX` if it is associated with a `UNIQUE` or `PRIMARY` constraint.
  - The behavior of DEFERRABLE `UNIQUE INDEX`.
    - Using INDEX and Trigger to check constraints.
    - Check each statement, not each row.
- `FOREIGN KEY` does not create `UNIQUE INDEX`, but only checks constraints with Trigger, so the behavior is the same between `NOT DEFERRABLE` and `DEFERRABLE`.

Reply:
Thanks for the detailed information.